### PR TITLE
Added Statistics logging and page for admins

### LINKED ISSUE #23: Add statistics
It would be cool to get a scope of the project. How many registered users do we have? How many request are made etc..
Maybe something for you @Tallerik ?

### PR DESCRIPTION
**Pull request by Tallerik**
## Dont merge now, just for code review.

### CODE REVIEW

File: src/statistics.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+<div id="visualize"></div>
+
+
+</section>
+<section>
+    <p>Anzahl registrierter Nutzer: <span id="registered"></span></p>
+
+</section>
+
+<script type="text/javascript" src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
```
Comment (Jeidnx) [inline src/statistics.html:30]: Just remove this line

File: src/data/statistics.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        var options = {
+            start: "2021-11-01",
+            end: getDate(),
+        };
+        var graph2d = new vis.Graph2d(container, dataset, options);
+
+
+    } catch (e) {console.log(e)}
+})
+
```
Comment (Jeidnx) [inline src/data/statistics.js:36]: I will (help) rewrite this to use XMLHTTPRequest instead of Jquery.

File: index.js [hunk 3]
```diff
@@ -319,3 +357,41 @@ app.get('*', (req, res) => {
...
+	return JSON.parse(d);
+}
+function saveData() {
+	fs.writeFile(file, JSON.stringify(stats), function (err) {
+		if(err)
+			console.log(err.message);
+		console.log("save")
+	});
+}
+function initScheduler() {
```
Comment (Jeidnx) [inline index.js:404]: maybe add the delay as a property so you can more easily change it in the head of the program

File: index.js [hunk 4]
```diff
@@ -319,3 +357,41 @@ app.get('*', (req, res) => {
...
+	setInterval(function () {
+		saveData();
+	}, 0.10 * 60 * 1000);
+}
+function getDate() {
+	const d = new Date();
+	return d.toISOString().slice(0, 10);
```
Comment (Jeidnx) [inline index.js:411]: just use 
`return new Date().toISOString()(...)`

File: index.js [hunk 5]
```diff
@@ -111,25 +111,49 @@ const classIdEnum = {
...
 app.get('/', (req, res) => {
+	const d = getDate();
+	if(!stats.requests.hasOwnProperty(d)) {
+		constructDateStruct(d)
+	}
+	stats.requests[d]["get"]["/"] += 1;
 	res.status(200).send(fs.readFileSync('./src/timetable.html', 'utf-8'));
 });
 app.get('/setup', (req, res) => {
+	const d = getDate();
```
Comment (Jeidnx) [inline index.js:153]: maybe date or curr or currentDate would be a better option for readability

File: index.js [hunk 6]
```diff
@@ -405,4 +430,13 @@ function createUserArray() {
...
 	return crypto.createHash("sha256").update(str).digest("hex");
+}
+function isUserAdmin(name) {
+	if(!config) {
+		return false;
+	}
+	if(!config.hasOwnProperty("adminuser")) {
+		return false;
+	}
+	return config.adminuser.includes(hash(name));
```
Comment (Jeidnx) [inline index.js:442]: suggest following:
```
function isUserAdmin(name) {
if(!config) {
    return false;
}
return config.adminuser?.includes(hash(name)) || false;
}

```
When adminuser is not set `config.adminuser?` will return `undefined`. `undefined  || false` will return false.

File: index.js [hunk 7]
```diff
@@ -113,6 +113,11 @@ const jwtSecret = process.env.JWT_SECRET;
 const schoolName = process.env.SCHOOL_NAME;
 const schoolDomain = process.env.SCHOOL_DOMAIN;
 const statsfileName = "data/data.json";
+let config = null;
+try {
+	config = require("./data/config.json");
+} catch (_) {}
+
```
Comment (Jeidnx) [inline index.js]: There has to be a better way for this

Comment (Tallerik) [reply index.js]: Only if the app is supposed to crash when config file not exists

**Review by Jeidnx — APPROVED** [review]

File: index.js (hunk 7 above)
Comment (Jeidnx) [reply index.js]: There you go:

```
const fs = require('fs');

let configPath = './config.json';
let config = null;
if (fs.existsSync(configPath)) {
	config = require(configPath);
}
console.log(config);
```

Comment (Tallerik) [reply index.js:116]: It just crashes now :smile:

**Pull request merged**